kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -245,272 +512,33 @@ object KinesisSink {
-  private val randomGenerator = new java.util.Random()
-
-  log.info("Creating thread pool of size " + kinesisConfig.threadPoolSize)
-  maybeSqs match {
-    case Some(sqs) =>
-      log.info(
-        s"SQS buffer for '$streamName' Kinesis sink is set up as: ${sqs.sqsBufferName}."
-      )
-    case None =>
-      log.warn(
-        s"No SQS buffer for surge protection set up (consider setting a SQS Buffer in config.hocon)."
-      )
-  }
-
-  implicit lazy val ec = concurrent.ExecutionContext.fromExecutorService(executorService)
 
   /**
-    * Recursively schedule a task to send everthing in EventStorage
-    * Even if the incoming event flow dries up, all stored events will eventually get sent
-    * Whenever TimeThreshold milliseconds have passed since the last call to flush, call flush.
-    * @param interval When to schedule the next flush
+    * Splits a Kinesis-sized batch of `Events` into smaller batches that meet the SQS limit.
+    * @param batch A batch of up to `KinesisLimit` that must be split into smaller batches.
+    * @param getByteSize How to get the size of a batch.
+    * @param maxRecords Max records for the smaller batches.
+    * @param maxBytes Max byte size for the smaller batches.
+    * @return A batch of smaller batches, each one of which meets the limits.
     */
-  def scheduleFlush(interval: Long = TimeThreshold): Unit = {
-    executorService.schedule(
-      new Thread {
-        override def run(): Unit = {
-          val lastFlushed = EventStorage.getLastFlushTime()
-          val currentTime = System.currentTimeMillis()
-          if (currentTime - lastFlushed >= TimeThreshold) {
-            EventStorage.flush()
-            scheduleFlush(TimeThreshold)
-          } else {
-            scheduleFlush(TimeThreshold + lastFlushed - currentTime)
-          }
-        }
-      },
-      interval,
-      MILLISECONDS
-    )
-    ()
-  }
-
-  // 'key' is used to populate the 'kinesisKey' message attribute for SQS
-  // and as partition key for Kinesis
-  case class Event(msg: ByteBuffer, key: String)
-
-  object EventStorage {
-    private var storedEvents              = List.empty[Event]
-    private var byteCount                 = 0L
-    @volatile private var lastFlushedTime = 0L
-
-    def store(event: Array[Byte], key: String): Unit = {
-      val eventBytes = ByteBuffer.wrap(event)
-      val eventSize  = eventBytes.capacity
-      if (eventSize >= MaxBytes) {
-        log.error(
-          s"Record of size $eventSize bytes is too large - must be less than $MaxBytes bytes"
-        )
-      } else {
-        synchronized {
-          storedEvents = Event(eventBytes, key) :: storedEvents
-          byteCount += eventSize
-          if (storedEvents.size >= RecordThreshold || byteCount >= ByteThreshold) {
-            flush()
-          }
-        }
+  def split(
+    batch: List[Events],
+    getByteSize: List[Events] => Int,
+    maxRecords: Int,
+    maxBytes: Int
+  ): List[List[Events]] = {
```

Comment:
First of all, you can rewrite this with `foldLeft` if you prefer.  It is roughly equivalent to what you wrote, so you can pick whichever you prefer:

```scala
 def split(
    batch: List[Events],
    getByteSize: List[Events] => Int,
    maxRecords: Int,
    maxBytes: Int
  ): List[List[Events]] =
batch.foldLeft(List.empty[List[Events]]) {
  case (Nil, next) => List(List(next))
  case (h :: t, next) if h.size + 1 > maxRecords || getByteSize(next :: h) > maxBytes =>
    List(next) :: h :: t
  case (h :: t, next) => (next :: h) :: t
}
```

Slightly more importantly.... you call `getBatchByteSize` for each event you add.  This is probably inefficient because `getBatchByteSize` allocates many byte buffers each time you call it.  It would be really nice if you can adapt the `foldLeft` so you accumulate the batch size at the same time as folding over the events.

Finally.... I think your `split` returns the events in reverse order.  Does this matter?  Do you un-reverse them before sending to sqs or is it unimportant?